src/analysis/cfg.h

```diff
@@ -77,8 +79,29 @@ struct CFG {
 
   void print(std::ostream& os, Module* wasm = nullptr) const;
 
+  // This must be done before getBlockIndex is called on an expression.
+  void computeExpressionBlockIndexes();
+
+  // Gets the index of the basic block in which the instruction resides.
+  // computeExpressionBlockIndexes() must be called first to populate the map.
+  Index getBlockIndex(Expression* expr) {
```

Comment:
Can we automatically call `computeExpressionBlockIndexes` here to make the API less stateful? Then we could make `computeExpressionBlockIndexes` private as well.

Reply:
We could, but always doing this computation would be significant extra work that is not needed in some (most?) cases. Personally I don't think requiring explicit opt-in to this feature makes the API noticeably worse.

Reply:
If we defer the computation until the first call to `getBlockIndex`, then we would only be doing the work when it is needed.

Reply:
That's an option. It does make the function no longer const, though, which is odd.

Reply:
Another option could be to add a friend class, `BlockIndexer` which would operate on a `CFG`. What do you think of that?

Reply:
> That's an option. It does make the function no longer const, though, which is odd.

This can be solved by making the index map `mutable`, but that's a gross hack.

> Another option could be to add a friend class, `BlockIndexer` which would operate on a `CFG`. What do you think of that?

This is my favorite option so far. I like the separation of concerns.

Reply:
Yeah, I guess even friendship isn't needed there. Done, PTAL.